### ISSUE TITLE
AWS is retiring the EC2-Classic environment

### ISSUE THREAD

**Issue by cgolubi1** [buttonmen-dev/buttonmen#2855]
AWS is retiring the environment called EC2-Classic, a legacy non-VPC environment, as of 2022-08-15.  We need to prepare for this by migrating all BM sites currently located in EC2-Classic, to VPC.

Status:
* Prod migration scheduled for 2022-08-07 10:00-10:30 EDT (14:00-14:30 UTC).

**Comment by cgolubi1** [comment]
The good news is that there should be no problem running BM in an instance within a VPC --- all of the replay sites i run in us-west-2 have been VPC from day one.

However, for historical reasons, all us-east-1 sites are in EC2-classic, and need to be migrated.  That means:
* prod and staging
* all dev sites, including the stats dev site

**Comment by cgolubi1** [comment]
Problems i'm aware of that we need to solve (note: i'm not saying these problems are hard, no need to spam me with proposed solutions --- i'm just trying to list everything we are not at-this-moment doing in an already-configured VPC, so that i have a todo list):
* We need to actually create a VPC for BM in us-east-1, and i need to make a vagrant config that uses it
* The replay sites aren't remotely accessible via web, only via SSH, so we have to configure a VPC to allow inbound traffic on 80 and 443
* The RDS databases for staging and prod (and the RDS test dev site, which we can use as a first example) need to be made available to in-VPC instances [TODO: check if the RDS instances themselves are in classic, or if they're safely in a VPC and only the connections need to move]
* The elastic IPs for staging and prod are non-VPC, and need to be either replaced with VPC ones and DNS updated (blech) or migrated to become VPC (preferable, but is it supported?)

**Comment by cgolubi1** [comment]
Order of operations:
1. create the us-east-1 VPC and stand up a fresh dev site in it
2. migrate the RDS test dev site to VPC, and make that list of steps
3. answer remaining TODOs, and make a plan for migrating staging and prod
4. migrate staging
5. schedule and migrate prod

**Comment by cgolubi1** [comment]
The PR addresses item 1 --- once that's merged, i'll go forward.

As we learned over in #2853, there are some bits of standing up a new RDS-backed instance that don't work correctly even absent VPC.  So i'll plan to put in a new PR to fix those as i migrate the RDS test dev site, so that when we get to standing up new staging and prod sites, there won't be an issue.

**Comment by cgolubi1** [comment]
Hmm, i think the simplest thing to do is to register an EIP for the RDS dev site, at least for now, because the IP identifying whether a site is RDS-backed is hardcoded into puppet config.  Obviously it'd be better to do this via instance tags the way we're doing the SSL stuff, but i don't want to bite that off right now.

So when i submit the RDS PR in a bit, it'll also contain an update to which IP address corresponds to the RDS dev site.

**Comment by cgolubi1** [comment]
Okay.  So i think the plan i summarized on #2857 is the correct set of steps for staging and prod (details will vary slightly from dev, and i'll outline those when i'm ready to go).  Once that PR is approved, i'll move forward with staging, and schedule prod.

Meanwhile, i shouldn't need any of those changes for the other existing dev sites, which are all local-DB-backed.  Here's a list of running dev sites currently in EC2 Classic:
* shadowshade's sandbox (stats site)
* tournaments dev site
* wildcard dev site
* tutorial dev site
* mad/mood dev site

I'm not going to bother to keep the mad/mood dev site, because people wanted the code to work very differently, and i understand the issue, so there's no value to that site at this moment.

I'll make a plan for the others now --- tutorial, wildcard, and *probably* tournaments, i can pretty easily stand up a clean dev site with the database from the existing dev site.  Let's go ahead and try to take care of those now.

**Comment by cgolubi1** [comment]
Okay, i've migrated tutorial, wildcard, and tournament --- no real issues; i had to update the tournaments database to apply `02709_pronouns.sql`, but that was easy enough.

**Comment by cgolubi1** [comment]
Shadowshade, now that we're at this point, let me know what you want me to do about the stats sandbox.  The issue is that the instance running that site will be turned off on 15 Aug.  Because the site is a snowflake (config was done by hand rather than within the site code), i can't just stand up a new copy.  And i know we want to get the site config puppetized, but i don't think there's time to be sure we can do that before 15 Aug.

The easiest thing for me to do is:
* stand up a new sandbox with fresh code, including restore whatever databases you tell me to onto that sandbox (`buttonmen-stats`?)
* give you a couple of weeks with both sandboxes running, so you can get your code working on the new sandbox
* On or before 15 Aug, teardown the old site and move the DNS name `stats.dev.buttonweavers.com` to point to the new sandbox

My suggestion is that i do that, and do it ASAP to give you as much time as possible.  Would something different be better for you?  If so, tell me what, and i'll think about whether it's feasible.

**Comment by blackshadowshade** [comment]
I'm currently only using the stats sandbox for storing the stats files, we're not actually using any functionality there.

I used it to demonstrate the possibility there of running R code on the site, and I noted that the major hurdle for the current code (which I'm happy to rework if necessary) is that there isn't enough memory on stats.dev to generate the huge matchup image.

I'm happy to upload our stats files to the main site instead if you want to get rid of stats.dev entirely. We need to decide what we're going to do longterm with running the stats code automatically, instead of me running it manually on my local machine and then uploading the files.

**Comment by cgolubi1** [comment]
Ah, okay!  In that case, for now, i'll just stand up a new site with the exact same databases as the existing one.  And, yes, let's get back to running the stats code automatically sooner rather than later --- i just want to be done with this instance deprecation situation before it's a crisis!

**Comment by cgolubi1** [comment]
Oh, and when you say "storing the stats files", you mean HTML files, not database contents at all.  So we don't actually even need to restore DBs.

**Comment by blackshadowshade** [comment]
That is correct.

**Comment by cgolubi1** [comment]
Okay, there you go --- migrated to new site, and hey now SSL works as a side effect of the code upgrade.  More about this soon, now that i actually understand what's going on with this site.

**Comment by cgolubi1** [comment]
LMK if anything is amiss, but since the functionality you're using is a DNS name and some static files, and it seems to be working for me, i'll assume it's good unless you tell me it's not.

**Comment by blackshadowshade** [comment]
Looks fine, the files at http://stats.dev.buttonweavers.com/ui/stats are accessible to me.

**Comment by cgolubi1** [comment]
Okay, so now all that's left is staging and prod --- i'll get back to this later this week, and cut over staging and make the plan to cut over prod.

The only complexity i haven't talked about is that now that we're in VPC, i don't really want to make another EBS volume to run backups --- i want to run backups to an EFS FS so i don't have to size it.  So i'll either have to snowflake the NFS client configuration or add it to the vagrant config.  More on that when i get there.

**Comment by cgolubi1** [comment]
Here's the list of all the steps i expect to need to migrate staging.buttonweavers.com to a VPC instance:

* Halt the old staging site, and move its vagrant files out of the way
* Launch the BM site
* As soon as possible, tag the instance (fqdn staging.buttonweavers.com in this case)
* As soon as possible, associate the staging EIP with the instance (this probably involves clicking a button to migrate the EIP to be a VPC EIP)
* Ctrl-C the vagrant provision when it's in "rsync"
* Run vagrant reload
* Run vagrant again for certbot/ssl:
```
vagrant rsync
vagrant provision
```

* Populate `/usr/local/etc/buttonmen_db_creds`
* Update the RDS DB security group to allow inbound access from the buttonmen-site security group
* At this point, `mysql_root_cli` should work
* Add a GRANT on the RDS DB for `bmuser1` at the in-VPC IP of the instance
* At this point, website access should work

I'll do all that, and record the end-to-end downtime of the site.  Assuming it's not truly ridiculous, i'll propose we just have the prod site be down for that length of time, on the theory that people would rather i just get this done and have bandwidth for other site work.

Note that this won't get the backups running, or rather, it'll get them running to `/`, which isn't a good place for them.  After this is done, i'll setup an EFS volume for staging's backups, and mount it via snowflaked commands (and then have a template to do the same thing for prod).

**Comment by cgolubi1** [comment]
Okay.  staging site should be back up --- total downtime 11m 2s.  I'm claiming that's good enough for prod.

**Comment by cgolubi1** [comment]
Scheduled prod migration for 2022-08-07 10:00-10:30 EDT.

**Comment by cgolubi1** [comment]
On-instance steps to attach backup volume:

```
sudo apt-get -y install nfs-common
sudo mount -t nfs4 ${FSID}.efs.us-east-1.amazonaws.com:/ /srv/backup
```

Add to `/etc/fstab`:

```
${FSID}.efs.us-east-1.amazonaws.com:/ /srv/backup nfs4 nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2,noresvport,_netdev 0 0
```

**Comment by blackshadowshade** [comment]
I agree, 12 minutes downtime should be perfectly fine.

**Comment by cgolubi1** [comment]
Okay!  My morning got away from me, but i'm starting now.

**Comment by cgolubi1** [comment]
Here's the list of all the steps i expect to need to migrate www.buttonweavers.com to a VPC instance:

* Halt the old www site, and move its vagrant files out of the way
* Launch the BM site (make sure it goes into us-east-1b)
* As soon as possible, tag the instance (fqdn www.buttonweavers.com in this case)
* As soon as possible, associate the prod EIP with the instance (this probably involves clicking a button to migrate the EIP to be a VPC EIP)
* Ctrl-C the vagrant provision when it's in "rsync"
* Run vagrant reload
* Run vagrant again for certbot/ssl:
```
vagrant rsync
vagrant provision
```

* Populate `/usr/local/etc/buttonmen_db_creds`
* Update the RDS DB security group to allow inbound access from the buttonmen-site security group
* At this point, `mysql_root_cli` should work
* Add a GRANT on the RDS DB for bmuser1 at the in-VPC IP of the instance
* At this point, website access should work

Then once that's done and the site is working, look at the backups.

**Comment by cgolubi1** [comment]
Okay, www.buttonweavers.com should be back up!  Let me know if anything is amiss.

**Comment by cgolubi1** [comment]
Okay, backups should also be enabled now, and i ran yesterday's to test.

**Comment by cgolubi1** [comment]
I think we can close this out now.
